pkg/planner/core/rule_predicate_simplification.go

```diff
@@ -221,6 +221,19 @@ func unsatisfiable(ctx base.PlanContext, p1, p2 expression.Expression) bool {
 	// Copy constant from equal predicate into other predicate.
 	equalValue := equalPred.(*expression.ScalarFunction)
 	otherValue := otherPred.(*expression.ScalarFunction)
+	// If equalValue's arg[1] and otherValue's arg[1] are string, we should get the right collation.
+	// See details at https://github.com/pingcap/tidb/issues/56479.
+	evalCtx := ctx.GetExprCtx().GetEvalCtx()
+	if equalValue.GetArgs()[1].GetType(evalCtx).EvalType() == types.ETString && otherValue.GetArgs()[1].GetType(evalCtx).EvalType() == types.ETString {
+		_, equalValueCollation := equalValue.CharsetAndCollation()
+		_, otherValueCollation := otherValue.CharsetAndCollation()
+		// If these two scalar functions have different collations, we don't want to compare them.
+		if equalValueCollation != otherValueCollation {
+			return false
+		}
+		equalValue.GetArgs()[1].GetType(evalCtx).SetCollate(equalValueCollation)
```

Comment:
why should we set it back? it has not changed?

Reply:
In the WHERE clause : WHERE c1 BETWEEN 'string1' AND 'string2' AND (c1 = 'string3' OR IsNull(c1)), 'string1' and 'string2' are set to the column's collation (utf8mb4_unicode_ci) when rewriting BETWEEN ... AND, but 'string3' is set to the connection-level collation (utf8mb4_general_ci). Before comparing "string1" and "string3" here, to avoid such collation mismatches, I explicitly reset the collations for the string constants.

Reply:
we have a `deriveCollation(ctx, funcName, args, retType, retType)` inside, which serves collation mismatch, while it's embedded in cast function, so i guess the better way here is to build a wrapper cast function as L237's new child but not for sure, you can have a try

Reply:
Thanks for the clarification! I’ll check out `deriveCollation` and try building a wrapper cast function like `BuildCastFunction` as you suggested.

Reply:
sorry, after discussion with @time-and-fate , seems we could use his suggested way, which will be more clear.